packages/rest/src/BaseResource.ts

```diff
@@ -0,0 +1,263 @@
+      // eslint-disable-next-line @typescript-eslint/no-this-alias
+      const resource: any = this;
+      cache[name] = construct().extend({
+        get name() {
+          return `${resource.name}.${name.replace(/#/, '')}`;
+        },
+      });
+    }
+    return cache[name].useFetchInit() as T;
+  }
+
+  /** Base endpoint that uses all the hooks provided by Resource  */
+  protected static endpoint(): RestEndpoint<
+    (this: RestEndpoint, params: any) => Promise<any>,
+    Schema | undefined,
+    undefined
+  > {
+    return this.memo('#endpoint', () => {
+      // eslint-disable-next-line
+      const resource = this;
+      const instanceFetch = this.fetch.bind(this);
+      const url = this.url.bind(this);
+
+      return new Endpoint(
+        function (params: any) {
+          return instanceFetch(this.url(params), this.getFetchInit());
+        },
+        {
+          ...this.getEndpointExtra(),
+          key: function (this: any, params: any) {
+            return `${this.method} ${this.url(params)}`;
+          },
+          url,
+          fetchInit: {} as RequestInit,
+          useFetchInit(this: any) {
+            this.fetchInit = resource.useFetchInit(this.fetchInit);
+            return this;
+          },
+          getFetchInit(this: any, body?: any) {
+            if (isPojo(body)) {
+              body = JSON.stringify(body);
+            }
+            return resource.getFetchInit({
+              ...this.fetchInit,
+              method: this.method,
+              signal: this.signal,
+              body,
+            });
+          },
+          method: 'GET',
+          signal: undefined as AbortSignal | undefined,
+        },
+      );
+    });
+  }
+
+  /** Base endpoint but for sideEffects */
+  protected static endpointMutate(): RestEndpoint<
+    (this: RestEndpoint, params: any, body?: any) => Promise<any>,
+    Schema | undefined,
+    true
+  > {
+    const instanceFetch = this.fetch.bind(this);
+    const endpoint = this.endpoint();
+    return this.memo('#endpointMutate', () =>
+      endpoint.extend({
+        fetch(this: RestEndpoint, params: any, body: any) {
+          return instanceFetch(this.url(params), this.getFetchInit(body));
+        },
+        sideEffect: true,
+        method: 'POST',
+      }),
+    );
+  }
+
+  static {
+    Object.defineProperty(this.prototype, 'url', {
+      get(): string {
+        // typescript thinks constructor is just a function
+        const Static = this.constructor as typeof BaseResource;
+        return Static.url(this);
+      },
+      set(v: string) {
+        Object.defineProperty(this, 'url', {
+          value: v,
+          writable: true,
+          enumerable: true,
+        });
+      },
+    });
+  }
+}
+
+const proto = Object.prototype;
+const gpo = Object.getPrototypeOf;
+
+function isPojo(obj: unknown): obj is Record<string, any> {
+  if (obj === null || typeof obj !== 'object') {
+    return false;
+  }
+  return gpo(obj) === proto;
+}
```

Comment:
`{ __proto__: null }` is a pojo also, but will fail this predicate

Reply:
how would that happen? seems an odd thing to pass to getFetchInit

Reply:
a null object? it's a normal thing to create if you want to avoid the possibility of key collisions.

Reply:
I found this when googling null object: https://stackoverflow.com/questions/801032/why-is-null-an-object-and-whats-the-difference-between-null-and-undefined

Can you find better docs or share example I'm still not understanding. Appreciate you!

Reply:
The term "null object" is just a colloquialism for "an object literal with a null [[Prototype]]", like `Object.create(null)`. The point of it is to remove the chance of collision with things on Object.prototype.

"pojo" checks are generally pretty weird - generally speaking, it's easier to have an argument either be a function, an array, or "we'll pretend it's an options bag".

Reply:
Maybe pojo is the wrong term? Really this is just to exclude other valid body arguments to fetch like strings, blobs, etc. 

Perhaps this is something you can create another PR for? As this code already exists and this PR is intended to just be a moving of code and updating of tests.

Reply:
that's up to you; i was just pointing out the problem.

Reply:
I do have `obj === null` above...and I get typeof undefined === 'undefined'. So it seems both those cases are handled